### DISCUSSION TITLE
Understanding SDFs' in Physx

### DISCUSSION THREAD

**Discussion by noblegmr** [NVIDIA-Omniverse/PhysX#298] [Q&A] [answered]
Hello Everyone!

I am currently working on a little project that involves doing collisions checking and I am trying to use the SDF capability that Physx has to offer. for context:

1. I have a static non-water tight mesh (STL), that I wish to calculate an SDF for (I know not ideal). 
2. I am generating a triangular mesh,  to create this SDF (attached the parameters DESC in an image to this post). following instructions from this [section](https://nvidia-omniverse.github.io/PhysX/physx/5.4.1/docs/Geometry.html#triangle-meshes), I am doing the generation on CPU since I am experiencing linker errors trying to include libPhysXGpu_64.so, created seperate discussion for this.
3. What I am doing next is query 3D points, using this SDF to get an approximate distance from surface. for which I am using this [function](https://nvidia-omniverse.github.io/PhysX/physx/5.4.1/_api_build/function_PxMathUtils_8h_1a26aee2af9614a783a6a65608b1e63f59.html?highlight=sdf%20sample#_CPPv411PxSdfSamplePK6PxRealRK6PxVec3RK6PxVec3RK6PxVec3K6PxRealK6PxRealK5PxU32K5PxU32K5PxU32R6PxVec36PxReal), which has little documentation attached to it.

I have been able to generate SDFs and query points successfully on some example water-tight and non-water tight meshes. The issues I am facing are:
- In understanding what this SDF is? : from a very preliminary standpoint and digging a little into the code I can see that it is some form of a voxel grid. 
- I've written a query function that utilises the sampleSDF function linked above, But it is not clear in terms of what some parameters are supposed to be in that function for example that is `sdfDx` and `invsdfDx`
- Digging into the source code I was able to come across different parameters for the SDF but these are not directly accessible? (inside GuSDF.h). I want the length of this SDF array that is generated so that I can store it (see next point).
- How I can save this SDF into a binary file? I tried to save my SDF into a binary file, load it in a different process and query the same points but I was unable to recreate the same results (encountered error discrepancies in order of 0.04 m ).
- Given a transform for the mesh, am I somehow able to transform my SDF after it has been generated as well? 

Attached here are two files that show how I am trying to setup and query an SDF. Any hints would be greatly appreciated!
[setup_SDFDesc.txt](https://github.com/user-attachments/files/16368690/setup_SDFDesc.txt)
[SDF_query_func.txt](https://github.com/user-attachments/files/16368692/SDF_query_func.txt)

Environment Information:
- Ubuntu 20.04, running inside a docker container with Physx 5.4.0.

**Comment by nvtw** [comment, marked as answer]
Hi, 
About your questions:
1. Non watertight meshes should work fine with SDFs. You could also use a free tool like MeshLab to easily convert them to wathertight meshes and use another simple format like .obj or .off.
2. I cannot tell where the linker errors come from. My assumption is a mistake in the CMake or any other build system you might be using.
3. It looks like PxSdfSample works only for dense SDFs and seems to be missing documentation. I'll add some documentation but that will take a while until it's reflected in the public github repository

- An SDF is a grid where each point stores the distance to the closest triangle on the triangle mesh. Positive distances mean that the smaple point is outside of the mesh, negative distances means the sample is inside.
- sdfDx is the cell size. invsdfDx = 1.0f/sdfDx. The cells are small cubes. Again it's bad that this function misses documentation.
- PhysX is not designed to make all data available in the public API but you might still find work arounds. But it's not officially supported to get the internal raw data of SDFs. However there is binary serialization support for the full triangle mesh which will store the SDF data as well. But nevertheless, for dense SDFs, the array length of the distance data is dimX * dimY * dimZ.
- An SDF should never be transformed since that is too expensive. You can compute the inverse of your transform and then use that to trans form your sample point from world to SDF space, evaluate the sdf and transform the results back to world space. That's how our collision solver works as well internally since that's much more efficient.

There is also a section about SDFs in the PhysX SDK documentation: https://nvidia-omniverse.github.io/PhysX/physx/5.4.1/docs/RigidBodyCollision.html#dynamic-triangle-meshes-with-sdfs
And there is also a page about serialization (including binary serialization): https://nvidia-omniverse.github.io/PhysX/physx/5.4.1/docs/Serialization.html

Hope those answers help a bit.

**Comment by AgentCodeBanks** [comment]
Hello,

I'm encountering an Access Violation exception in GPU memory when deserializing certain TriangleMesh assets that contain Signed Distance Fields (SDFs). This happens after I have successfully serialized the mesh to a binary file. While this process works for some meshes, others consistently cause an exception on reload.

I am using the binary serialization process for TriangleMeshes as outlined in the PhysX 5.4.1 documentation: https://nvidia-omniverse.github.io/PhysX/physx/5.4.1/docs/Serialization.html.

My question is: How is the GPU memory allocated during the deserialization of the SDF, and what could cause an Access Violation in this process?

Any guidance on how to debug this exception would be greatly appreciated.
